pkg/auth/oauth2/github.go

```diff
@@ -0,0 +1,187 @@
+func (g *GitHubAuthenticator) RequestAuth() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		url := g.config.GoOAuth2Config.AuthCodeURL(g.state, oauth2.AccessTypeOffline)
+		http.Redirect(w, r, url, http.StatusFound)
+	}
+}
+
+// Callback returns a http.HandlerFunc that exchanges the authorization code for an access token and fetches user info from the OAuth2 provider.
+// The parameter is a function that takes the user info and returns the JWT token or an error.
+func (g *GitHubAuthenticator) Callback(loginAndGetJWT func(userInfo any) (token *jwt.TokenDetails, err errors.Error)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		code := r.URL.Query().Get(codeParam)
+		state := r.URL.Query().Get(stateParam)
+		if state != g.state {
+			g.lc.Error("State does not match, you may be under CSRF attack.")
+			http.Error(w, "invalid state. You may be under CSRF attack.", http.StatusUnauthorized)
+			return
+		}
+
+		token, err := g.config.GoOAuth2Config.Exchange(r.Context(), code)
+		g.lc.Debugf("exchange authentication code %v for the access token", code)
+		if err != nil {
+			g.lc.Errorf("failed to exchange token, err: %v", err)
+			http.Error(w, fmt.Sprintf("failed to exchange token, err: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		client := g.config.GoOAuth2Config.Client(r.Context(), token)
+		resp, err := client.Get(g.config.UserInfoURL)
+		g.lc.Debugf("fetching user info from %v", g.config.UserInfoURL)
+		if err != nil {
+			g.lc.Errorf("failed to fetch user info: %v", err)
+			http.Error(w, fmt.Sprintf("failed to fetch user info: %v", err), http.StatusInternalServerError)
+			return
+		}
+		defer resp.Body.Close()
+
+		userData, err := io.ReadAll(resp.Body)
+		if err != nil {
+			g.lc.Errorf("failed to read the response body: %v", err)
+			http.Error(w, fmt.Sprintf("fail to read the response body: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		userInfo := GitHubUserInfo{}
+		err = json.Unmarshal(userData, &userInfo)
+		if err != nil {
+			g.lc.Errorf("failed to parse the response body: %v", err)
+			http.Error(w, fmt.Sprintf("fail to parse the response body: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		// Store the token details in the map
+		strID := strconv.FormatInt(userInfo.ID, 10)
+		g.mu.Lock()
+		g.tokens[strID] = token
+		g.mu.Unlock()
+
+		tokenDetails, err := loginAndGetJWT(userInfo)
+		if err != nil {
+			g.lc.Errorf("failed to log in: %v", err)
+			http.Error(w, fmt.Sprintf("failed to log in: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		// Set the tokens to cookie and redirect to the redirect path
+		jwt.SetTokensToCookie(w, tokenDetails)
+		http.Redirect(w, r, g.config.RedirectPath, http.StatusSeeOther)
+	}
+}
+
+// GetTokenByUserID returns the oauth2 token by user ID
+func (g *GitHubAuthenticator) GetTokenByUserID(userId string) (*oauth2.Token, errors.Error) {
```

Comment:
tried to use composition to write the common functions in a `baseOAuthAuthenticator`
https://golangbot.com/inheritance/